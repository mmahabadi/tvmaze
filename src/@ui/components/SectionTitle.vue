<script setup lang="ts">
defineProps<{ title: string }>()
</script>

<template>
  <div class="my-9">
    <div class="w-full flex sm:gap-8 gap-4 items-center">
      <div class="text-teal-500"><slot></slot></div>
      <h2 class="sm:text-xl font-bold text-lg">{{ title }}</h2>
    </div>
  </div>
</template>
