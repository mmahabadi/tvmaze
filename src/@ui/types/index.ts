export type SwiperOptions = {
  centeredSlides: boolean
  slidesPerView: number
  spaceBetween: number
  navigation: boolean
  loop: boolean
  scrollbar: { draggable: boolean }
  speed: number
  lazy: boolean
  autoplay: {
    delay: number
    disableOnInteraction: boolean
  }
}
