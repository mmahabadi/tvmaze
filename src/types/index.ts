export type AppSettings = {
  itemsInEachGenreSection: number
  itemsInEachPage: number
  itemsToShowInHomePageSlider: number
  yearLimitToConsiderAsRecentToShowInHomePageSlider: number
}

export type Rating = {
  average: number
}

export type Schedule = {
  time: string
  days: string[]
}

export type Image = {
  medium: string
  original: string
}

export type ShowByGenre = {
  genre: string
  movies: Movie[]
  loading: boolean
  error: null | Error
}

export type Movie = {
  id: number
  name: string
  type: string
  language: string
  genres: string[]
  status: string
  ended: string
  officialSite: string
  schedule: Schedule
  rating: Rating
  image: Image
  summary: string
  _embedded: Embedded
}

export type Embedded = {
  cast: Cast[]
}

export type Cast = {
  person: Person
  character: Character
}

export type Person = {
  id: number
  url: string
  name: string
  country: Country
  birthday: string
  deathday: string | null
  image: Image
  gender: string
}

export type Country = {
  name: string
  code: string
  timezone: string
}

export type Character = {
  id: number
  url: string
  name: string
  image: Image
}

export type SlideItem = Pick<Movie, 'id' | 'name'> & { rating: number; image: string }

export type CastItemProps = Pick<Person, 'name'> & { image: string }
