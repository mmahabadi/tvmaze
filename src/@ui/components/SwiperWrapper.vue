<script setup lang="ts">
import { Swiper } from 'swiper/vue'
import { Autoplay, Pagination, Navigation } from 'swiper/modules'
import { type SwiperOptions } from '@/@ui/types'
import { defineProps, computed } from 'vue'

import 'swiper/css'
import 'swiper/css/autoplay'
import 'swiper/css/navigation'
import 'swiper/css/pagination'
import 'swiper/css/scrollbar'

const props = defineProps<Partial<SwiperOptions>>()

const swiperStyle = {
  '--swiper-navigation-color': 'rgb(20 184 166 / var(--tw-text-opacity))',
  '--swiper-pagination-color': 'rgb(20 184 166 / var(--tw-text-opacity))'
}
const swiperOptions = computed(() => ({
  ...props
}))

const convertToInt = (prop: number | undefined) => {
  return +(prop ?? 0)
}

const breakpoints = computed(() => ({
  0: {
    slidesPerView: Math.min(1, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(10, convertToInt(props.spaceBetween))
  },
  768: {
    slidesPerView: Math.min(2, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(10, convertToInt(props.spaceBetween))
  },
  1024: {
    slidesPerView: Math.min(3, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(20, convertToInt(props.spaceBetween))
  },
  1280: {
    slidesPerView: Math.min(5, convertToInt(props.slidesPerView)),
    spaceBetween: Math.min(30, convertToInt(props.spaceBetween))
  },
  1536: {
    slidesPerView: convertToInt(props.slidesPerView),
    spaceBetween: convertToInt(props.spaceBetween)
  }
}))
</script>
<template>
  <swiper
    :style="swiperStyle"
    :modules="[Autoplay, Pagination, Navigation]"
    v-bind="swiperOptions"
    :breakpoints="breakpoints"
  >
    <slot></slot>
  </swiper>
</template>
