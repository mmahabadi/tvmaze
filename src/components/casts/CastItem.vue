<script setup lang="ts">
import { type CastItemProps } from '@/types'
import Image from '@/@ui/components/Image.vue'

defineProps<CastItemProps>()
</script>
<template>
  <div class="cast-item">
    <Image :src="image" :alt="name" class="img" />
    <p class="text-center">{{ name }}</p>
  </div>
</template>
<style scoped>
.cast-item {
  @apply w-full p-3 rounded flex-col bg-gray-900 border border-gray-800;

  .img {
    @apply w-full h-96 object-cover rounded mb-4;
  }
}
</style>
