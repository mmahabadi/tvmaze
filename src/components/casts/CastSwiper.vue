<script setup lang="ts">
import { SwiperSlide } from 'swiper/vue'
import SwiperWrapper from '@/@ui/components/SwiperWrapper.vue'
import { type Person } from '@/types'
import { type SwiperOptions } from '@/@ui/types'
import CastItem from './CastItem.vue'
import UserIcon from '@/@ui/components/icons/UserIcon.vue'
import SectionTitle from '@/@ui/components/SectionTitle.vue'

defineProps<{ slides: Person[] }>()

const options = {
  slidesPerView: 4,
  spaceBetween: 50,
  navigation: true,
  autoplay: {
    delay: 5000
  }
} as SwiperOptions
</script>
<template>
  <div class="mb-12">
    <SectionTitle title="Casts">
      <UserIcon />
    </SectionTitle>
    <SwiperWrapper v-bind="options">
      <swiper-slide v-for="slide in slides" :key="slide.id">
        <CastItem :item="slide" />
      </swiper-slide>
    </SwiperWrapper>
  </div>
</template>
