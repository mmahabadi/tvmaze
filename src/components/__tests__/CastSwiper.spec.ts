import { describe, it, expect, beforeEach } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import CastSwiper from '../casts/CastSwiper.vue'
import SwiperSlide from 'swiper/vue'
import SwiperWrapper from '@/@ui/components/SwiperWrapper.vue'
import CastItem from '../casts/CastItem.vue'
import UsersIcon from '@/@ui/components/icons/UsersIcon.vue'
import SectionTitle from '@/@ui/components/SectionTitle.vue'
import { type Image, type Person } from '@/types'

describe('CastSwiper.vue', () => {
  let wrapper: VueWrapper<unknown>
  const props = {
    slides: [
      {
        id: 1,
        name: 'Person 1',
        image: { medium: 'image1.jpg', original: 'image1.jpg' } as Image,
        url: ' ',
        country: ' ',
        birthday: '',
        gender: '',
        deathday: null
      } as unknown as Person
    ]
  }

  beforeEach(() => {
    wrapper = mount(CastSwiper, {
      props,
      global: {
        components: {
          SwiperSlide,
          SwiperWrapper,
          CastItem,
          UsersIcon,
          SectionTitle
        }
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the correct number of slides', () => {
    const slides = wrapper.findAllComponents(CastItem)
    expect(slides.length).toBe(props.slides.length)
  })
})
