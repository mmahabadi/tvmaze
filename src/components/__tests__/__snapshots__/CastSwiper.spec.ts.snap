// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CastSwiper.vue > renders correctly 1`] = `
"<div class="mb-12">
  <div class="my-9">
    <div class="w-full flex sm:gap-8 gap-4 items-center">
      <div class="text-teal-500"><svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 640 512" class="sm:w-6 sm:h-6 w-4 h-4 text-subMain" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg">
          <path d="M192 256c61.9 0 112-50.1 112-112S253.9 32 192 32 80 82.1 80 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C51.6 288 0 339.6 0 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zM480 256c53 0 96-43 96-96s-43-96-96-96-96 43-96 96 43 96 96 96zm48 32h-3.8c-13.9 4.8-28.6 8-44.2 8s-30.3-3.2-44.2-8H432c-20.4 0-39.2 5.9-55.7 15.4 24.4 26.3 39.7 61.2 39.7 99.8v38.4c0 2.2-.5 4.3-.6 6.4H592c26.5 0 48-21.5 48-48 0-61.9-50.1-112-112-112z"></path>
        </svg></div>
      <h2 class="sm:text-xl font-bold text-lg">Casts</h2>
    </div>
  </div>
  <div class="swiper swiper-initialized swiper-horizontal" style="--swiper-navigation-color: rgb(20 184 166 / var(--tw-text-opacity)); --swiper-pagination-color: rgb(20 184 166 / var(--tw-text-opacity));" lazy="false">
    <div class="swiper-wrapper" style="transition-duration: 0ms; transition-delay: 0ms; transform: translate3d(NaNpx, 0px, 0px);">
      <div class="swiper-slide">
        <div data-v-6563c4e7="" class="cast-item"><img data-v-6563c4e7="" src="image1.jpg" alt="Person 1" class="w-full h-full object-cover img">
          <p data-v-6563c4e7="" class="text-center">Person 1</p>
        </div>
        <!---->
      </div>
    </div>
    <div class="swiper-button-prev"></div>
    <div class="swiper-button-next"></div>
    <!---->
    <!---->
  </div>
</div>"
`;
